internal/jobtracker/jobtracker.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2025 Canonical.
+
+// Package jobtracker provides a way to run routines in one instance of JIMM and track them in another.
+// That is, their status can be checked and they can be stopped.
+
+package jobtracker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/juju/clock"
+	"github.com/juju/retry"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+)
+
+// JobTrackerStore defines the interface for tracking the lifecycle and status of jobs.
+// It provides methods to add a new job, update its status (running, successful, or failed),
+// and retrieve a stop signal for a specific job.
+type JobTrackerStore interface {
+	AddJob(ctx context.Context, jobType string) (uuid.UUID, error)
+	SetJobRunning(ctx context.Context, jobId uuid.UUID) error
+	SetJobSuccessful(ctx context.Context, jobId uuid.UUID) error
+	SetJobFailed(ctx context.Context, jobId uuid.UUID, jobErr error) error
+	GetJobStopSignal(ctx context.Context, jobId uuid.UUID) (stopSignal bool, err error)
+}
+
+// Tracker manages job tracking operations using a provided JobTrackerStore.
+// It periodically performs tasks based on the specified stopInterval duration.
+type Tracker struct {
+	store        JobTrackerStore
+	stopInterval time.Duration
+}
+
+// NewJobTracker creates and returns a new Tracker instance using the provided JobTrackerStore and stopInterval.
+// It returns an error if the store is nil or if stopInterval is not greater than zero.
+func NewJobTracker(store JobTrackerStore, stopInterval time.Duration) (*Tracker, error) {
+	tracker := &Tracker{}
+	if store == nil {
+		return tracker, errors.New("store cannot be nil")
+	}
+	if stopInterval <= 0 {
+		return tracker, errors.New("stopInterval must be greater than zero")
+	}
+	tracker.stopInterval = stopInterval
+	tracker.store = store
+
+	return tracker, nil
+}
+
+// Run runs a new job and reurns the job ID.
+func (j *Tracker) Run(ctx context.Context, jobType string, job func(ctx context.Context) error, deadline time.Duration) (uuid.UUID, error) {
+	jobId, err := j.store.AddJob(ctx, jobType)
+	if err != nil {
+		return jobId, err
+	}
+
+	go j.runJob(ctx, jobId, deadline, job)
```

Comment:
Need to come back to this after lunch, but one thing that stood out to me, is that I think we want a new context when starting the job. The context associated with the call to `Run` will end once the user's request completes.

Reply:
We create a new one within runJob but use the top level context as a parent. I guess we could just remove the parent if you're worried people might cancel it?